### PR TITLE
✅🏗 Ban chai should assertions

### PR DESCRIPTION
**Pull request by jridgewell**
Stage 1 of a multi part cleanup of our test suites. Unlovingly called "i really really really hate chai", because it's **far** to easy to write broken tests. These tests aren't bad themselves, but they way chai allows you to write certain assertions leads to bugs. The fault is with chai's format, not with the tests themselves.

The main reason is that I'm writing several autofixers, and I need to convert the codebase to `expect` syntax so that its easier to find all cases.

But, using `should` is kinda bad. For it to work, it has to put a non-enumerable getter on `Object.prototype` (always a bad idea to monkey with builtin prototypes). And it's inconsistent, try doing `foo.should.be.ok` when `foo` is `undefined` or `Object.create(null)`.

Current roadmap:
- Ban `should` usage
- Ban chai-as-promised
- Lint for chai assertion methods that are not called.

### CODE REVIEW

File: build-system/eslint-rules/i-really-really-really-dislike-chai.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    fixable: 'code',
+  },
+
+  create(context) {
+    return {
+      'MemberExpression[property.name=should]': function (node) {
+        // Ban using .should assertions.
+        context.report({
+          node,
+          message: 'Use expect() assertions only',
```
Comment (rsimha) [inline build-system/eslint-rules/i-really-really-really-dislike-chai.js:29]: Can you add some context to the PR description about why this rule was necessary?

**Comment by jridgewell** [comment]
> I genuinely share your frustration about badly written tests.

I should rephrase, my snark isn't appropriate. These tests aren't bad themselves, but they way chai allows you to write certain assertions leads to bugs. The fault is with chai's format, not with the tests themselves.

> However, I’m not a fan of repeating the word “really” for style reasons

Lol, this is meant to be read like https://genius.com/amp/Carly-rae-jepsen-i-really-like-you-lyrics.

> Perhaps I can convince you to rename this rule `no-should-assertions`

There'll be about 4 PRs to cleanup chai's bad parts. I guess I can group them with a common "avoid-bad-chai-FOO".

**Comment by choumx** [comment]
I think I'm missing some context or offline conversation. Why is `should` bad?

**Comment by jridgewell** [comment]
> Can you add some context to the PR description about why this rule was necessary?
> I think I'm missing some context or offline conversation. Why is should bad?

The main reason is that I'm writing several autofixers, and I need to convert the codebase to `expect` syntax so that its easier to find all cases.

But, using `should` is kinda bad. For it to work, it has to put a non-enumerable getter on `Object.prototype` (always a bad idea to monkey with builtin prototypes). And it's inconsistent, try doing `foo.should.be.ok` when `foo` is `undefined` or `Object.create(null)`.

**Comment by choumx** [comment]
Thanks but I think I'm still missing the complete picture here. Can you elaborate on what autofixers you're planning on building and how `should` and `chai-as-promised` are? 

And we should have some discussion on impact (how should devs write tests with minimal chai?) and any alternatives. This might be the right move but we probably shouldn't make it hastily or unilaterally.

**Comment by jridgewell** [comment]
> Can you elaborate on what autofixers you're planning on building and how should and chai-as-promised are blocking it?

`should` blocks it because if I allow `foo.should.eventually.be.fullfiled`, I now have to fix that _and_ `expect(foo).to.be.fulled` formats. The overwhelming majority of the codebase uses `expect`, so standardizing on it simplifies everything. Plus we get rid of these minor issues with `should`.

As for `chai-as-promised`, it's makes it extremely easy to create out-of-band assertions. Eg, https://github.com/ampproject/amphtml/blob/d7480aad08f503ff9c8f271b59af99e67175ac78/extensions/amp-ad-exit/0.1/test/test-amp-ad-exit.js#L752-L754.

`chai-as-promised` essentially works by chaining a `.then` to the value, and returning that new promise chain. If you don't wait for that return value, we have a bug. I've found quite a few that fall into this same trap.

But it's not needed anymore. Now that we allow async await, it'd be much easier to use that to write tests: https://astexplorer.net/#/gist/f1fec5455f5436b2a55757d05406f188/4339b244925955f447421956554e480a44d7b731

> And we should have some discussion on impact (how should devs write tests with minimal chai?) and any alternatives.

This is all prompted by the ease with which chai lets you write invalid assertions. Eg, https://github.com/ampproject/amphtml/blame/8bbcbc79119bd5137594d22ea22f6f5a9376bb79/test/unit/test-dom.js#L1102, which doesn't actually assert anything because it's a method and must be called.

Simplifying the way that assertions can be made will make this much easier to lint for. My ideal goal (but something that will come much later) it to replace chai entirely with a compatible test library that requires you to call all assertions (the only difference), so we never get this problem again.

**Comment by choumx** [comment]
I'm not in love with chai's API either, but let's continue this in design review next week? And get some more voices into the discussion.

**Comment by jridgewell** [comment]
I'd be happy to have a design review about removing `chai-as-promised`.  We can either continue using it, and add more linting around improper usage, or we can remove it and take advantage of async/await. But removing `chai-as-promised` is unrelated to this PR.

But I don't think this PR should be blocked on that. There are currently 144 `.should`s, and 26,782 `expect()`s. Standardizing on `expect()` now allows me to work on the actual linting, so that we can find all of the not-called assertion methods.

**Comment by choumx** [comment]
> But removing chai-as-promised is unrelated to this PR.

Oh, I read this as "step 1" of a larger change per the roadmap in the PR description. 

Also, it's still not clear to me exactly what autofix/linting you want to do. :) Check that chai functions are not misused as properties? Check that Promise assertions are returned or `await`'ed? Anything else?

I think what can help bring some clarity is by filing an issue with (1) a clear statement of the problem with examples (you already gave a few above), (2) your recommended solution, and (3) alternatives with some tradeoff analysis. 

Then a linked PR (this) would make a lot more sense and readers wouldn't have to dig for context. I appreciate the initiative/bias for action here but I think some process is warranted here.

**Pull request closed without merging**